File: backend/app/routers/historical_impacts.py
from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session

from app.database import get_db
from app import models

router = APIRouter(
    prefix="/historical-impacts",
    tags=["historical impacts"],
)


@router.get("/")
def get_historical_impacts(
    stakeholder_type: str | None = None,
    impact_type: str | None = None,
    db: Session = Depends(get_db),
):
    query = db.query(models.HistoricalImpact)

    if stakeholder_type:
        query = query.filter(
            models.HistoricalImpact.stakeholder_type
            == stakeholder_type
        )

    if impact_type:
        query = query.filter(
            models.HistoricalImpact.impact_type
            == impact_type
        )

    impacts = query.all()

    result = []

    for impact in impacts:
        semester = (
            db.query(models.Semester)
            .filter(models.Semester.id == impact.semester_id)
            .first()
        )

        stakeholder_name = impact.stakeholder_id

        if impact.stakeholder_type == "lecturer":
            lecturer = (
                db.query(models.Lecturer)
                .filter(models.Lecturer.id == impact.stakeholder_id)
                .first()
            )

            if lecturer:
                stakeholder_name = lecturer.name

        elif impact.stakeholder_type == "cohort":
            cohort = (
                db.query(models.Cohort)
                .filter(models.Cohort.id == impact.stakeholder_id)
                .first()
            )

            if cohort:
                stakeholder_name = cohort.name

        result.append(
            {
                "id": impact.id,

                "semester_id": impact.semester_id,
                "semester_name": (
                    semester.name
                    if semester
                    else impact.semester_id
                ),

                "stakeholder_type": impact.stakeholder_type,
                "stakeholder_id": impact.stakeholder_id,
                "stakeholder_name": stakeholder_name,

                "constraint_id": impact.constraint_id,
                "impact_type": impact.impact_type,

                "occurred_on": impact.occurred_on,
                "day": impact.day,

                "magnitude_minutes": impact.magnitude_minutes,
                "details": impact.details,
            }
        )

    return result

File: frontend/src/features/timetable/components/solution/historicalImpacts.ts
export type HistoricalStakeholderType =
  | "lecturer"
  | "cohort";

export type HistoricalImpactType =
  | "lunch-break-reduced"
  | "consecutive-teaching";

export type HistoricalImpactDetails = {
  expected_minutes?: number;
  received_minutes?: number;
  limit_minutes?: number;
  actual_minutes?: number;
};

export type HistoricalImpact = {
  id: string;

  semester_id: string;
  semester_name: string;

  stakeholder_type: HistoricalStakeholderType;
  stakeholder_id: string;
  stakeholder_name: string;

  constraint_id: string;
  impact_type: HistoricalImpactType;

  occurred_on: string;
  day: string;

  magnitude_minutes: number;

  details: HistoricalImpactDetails;
};

export async function fetchHistoricalImpacts(
  stakeholderType: HistoricalStakeholderType,
  impactType: HistoricalImpactType,
): Promise<HistoricalImpact[]> {
  const params = new URLSearchParams({
    stakeholder_type: stakeholderType,
    impact_type: impactType,
  });

  const response = await fetch(
    `http://127.0.0.1:8000/historical-impacts/?${params.toString()}`,
  );

  if (!response.ok) {
    throw new Error(
      `Failed to fetch historical impacts: ${response.status}`,
    );
  }

  return response.json();
}